Make worker block lock pool size observable

elega opened the pull request:
### What changes are proposed in this pull request?

This PR implements a block lock resource pool for workers with more logs and metrics added.

### Why are the changes needed?

The pool size should be observable by metric and logging, so we know when this pool is full.

### Does this PR introduce any user facing changes?

N/A

Implemented #14803

HelloHorizon commented:
@beinan Can you take a look?

beinan commented on core/common/src/main/java/alluxio/metrics/MetricKey.java:
```diff
@@ -1406,6 +1406,12 @@ public MetricKey build() {
               + "Use this metric to monitor the RPC pressure on worker.")
           .setMetricType(MetricType.GAUGE)
           .build();
+  public static final MetricKey WORKER_BLOCK_LOCK_POOL_REMAINING_RESOURCES_KEY =
```
To be consistent with other metric keys
```suggestion
  public static final MetricKey WORKER_BLOCK_LOCK_POOL_REMAINING_RESOURCES_COUNT =
```

elega replied:
sure thanks for flagging this.

beinan commented on core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java:
```diff
@@ -0,0 +1,105 @@
+
+    MetricsSystem.registerGaugeIfAbsent(
+        MetricKey.WORKER_BLOCK_LOCK_POOL_REMAINING_RESOURCES_KEY.toString(),
+        this::getRemainingPoolResources
```
does "mRemainingPoolResources::get" work here?

beinan replied:
I'm just thinking if we could just watch the size of `mResources` from the super class?

elega replied:
1. yes it worked. made the change already.
2. unfortunately that doesn't work. Basically the pool creates resources by lazy and the size of mResource is only the number of "created" resources. For example the pool size is 10 and a thread acquires a lock and releases it later. the size of mResource is 1 but the size of remaining pool resources should be 10.

beinan commented on core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java:
```diff
@@ -0,0 +1,105 @@
+        && !mHasReachedFullCapacity.get()
+        && mHasReachedFullCapacity.compareAndSet(false, true)
+    ) {
+      LOG.warn("Block lock manager client RW lock pool exhausted. Capacity size {}", mMaxCapacity);
```
do we print this log before?  my concern is it might print too much logs on production

elega replied:
haha i don't think this is a problem because i used atomic boolean CAS to make sure the log is only printed ONCE on the first time the pool reaches full.

beinan replied:
I see, super smart.  But when will you reset the mHasReachedFullCapacity? if you never reset this flag, what's the purpose of printing this log?  Alluxio worker is a long run service,  resource exhaustion can happen at any time, but we won't see this log except the first time it happens.

elega replied:
okay switched to use SAMPLING_LOG.

beinan commented on core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  @Override
+  public ClientRWLock acquire() {
```
do we really need this override method?

elega replied:
yes. 
this is 
**this**.acquire(WAIT_INDEFINITELY, null);
instead of 
**super**.acquire(WAIT_INDEFINITELY, null);

beinan replied:
I don't write much java, but I think this override is redundant, we could safely remove it, and this.acquire would still be invoked.  I think the bytecode of this invocation is `invokeDynamic` rather than `invokeStatic`

jiacheliu3 replied:
I think beinan is right. If not implemented, `BlockRWLockPool` will inherit the parent behavior and invoke `this.acquire(time, unit)` implicitly. 

I never expected to discuss bytecode with you guys when I clicked on this PR LOL. I guess the quickest way to verify is to test in the UT.

elega replied:
yeah i was wrong... sorry for the confusion.

beinan reviewed (commented):
Nice work @elega !  I just put a few comments, thx for the contribution!

elega commented:
@beinan hi thanks for reviewing my first PR on alluxio. I addressed some of your comments and replied the rest. Would you mind taking a look when you have time? Thanks a lot!

beinan commented on core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java:
```diff
@@ -0,0 +1,97 @@
+
+  @Override
+  public void release(ClientRWLock resource) {
+    mRemainingPoolResources.incrementAndGet();
```
Good catch, but what if the release function failed somehow?

jiacheliu3 replied:
Good question @beinan but the `release()` does not have checked exceptions. I don't see what unchecked exceptions can be thrown at the moment. If the unchecked exceptions are really thrown, will the worker just exit(so counters don't matter no more)?

elega replied:
yeah i feel if the aquire/release functions throws exceptions explicitly, it's a good idea to take them into consideration and try catch them. But in this case, it's quite clear that if any exception is thrown by a lock pool, the worker won't be able to handle such case and exiting seems to be the only outcome of that.

jiacheliu3 commented on core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java:
```diff
@@ -0,0 +1,105 @@
+  @Override
+  public ClientRWLock acquire(long time, TimeUnit unit) {
+    ClientRWLock ret = super.acquire(time, unit);
+    maybeLogPoolReachesFullCapacity();
```
If the full capacity has been reached, I'd actually expect to see some log before getting blocked and wait. So when I see my application stuck somewhere, I can tell from the log why I'm stuck. Therefore printing before and after acquire act slightly differently.

**Print after acquire:**
When the one request takes the last resource, a message will be logged. If no request releases, and some time elapses, and my request comes in and get blocked, it may not be obvious why my request blocks because the previous log can be far ahead in the log.

**Print before acquire:**
If the resource are already out, I can see a message logged right there matching the timestamp of my request. It's not perfect but slightly more obvious.

There's always a gap between the check(log) and `acquire()` so the logs are not 100% precious. Maybe an even better way is to NEVER `acquire()` indefinitely. When the `TimeUnit` is null or the timeout is pretty long, we use a while loop to acquire with smaller timeouts but track the total time elapsed instead. Then the logging is much much more flexible.

elega replied:
yup i totally agree with this. 
Actually I have come up with a better solution. For the acquire() function without any args, instead of slicing the lock acquisition process, we can simply do a timed lock acquisition first and give it enough long time, say 1 second. If the lock acquisition failed, we print a log about the thread pool depletion. Then we can turn into the unlimited waiting mode to acquire the lock.

jiacheliu3 reviewed (changes requested):
Mostly LGTM despite a few ongoing minor discussions. Thanks for the contribution!

elega commented on core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java:
```diff
@@ -0,0 +1,106 @@
+   */
+  @Override
+  public ClientRWLock acquire() {
+    ClientRWLock ret = this.acquire(1, TimeUnit.SECONDS);
```
@jiacheliu3  please review this change. ^^

jiacheliu3 replied:
This approach LGTM. The only open question is, what is a good time to wait? 1 sec is fine for me. @beinan WDYT?

IMO I don't think we need a property key to configure this time, it is too trivial and it'll be really hard for users to understand what it does.

jiacheliu3 commented on core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java:
```diff
@@ -0,0 +1,106 @@
+@ThreadSafe
+public class BlockRWLockPool extends ResourcePool<ClientRWLock> {
+  private static final Logger SAMPLING_LOG =
+      new SamplingLogger(LoggerFactory.getLogger(BlockRWLockPool.class),
+          30L * Constants.SECOND_MS);
```
I think we can totally use a much larger interval, like 10 seconds. I figure the point of logging multiple times is to make sure the user can see the log message enough times to catch his attention when he scrolls up and down a little bit in the log. More messages than that will be unnecessary.

jiacheliu3 commented on core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java:
```diff
@@ -0,0 +1,106 @@
+      SAMPLING_LOG.error(
+          "Lock acquisition failed due to lock pool resources being exhausted. "
+              + "Pool capacity size {}; lock acquisition duration {}{}",
+          mMaxCapacity,
+          time,
+          unit
```
nit: We rarely separate each param into one line, maybe you can chug those params a little to find your comfortable balance. We don't have an explicit rule of how many params can or cannot be on the same line.

jiacheliu3 commented on core/server/worker/src/test/java/alluxio/worker/block/BlockRWLockPoolTest.java:
```diff
@@ -0,0 +1,123 @@
+          new CyclicBarrier(numLocksEarlyRelease + numLocksLateRelease + 1);
+      final CyclicBarrier barrier2 = new CyclicBarrier(numLocksLateRelease + 1);
+      for (int j = 0; j < numLocksEarlyRelease; ++j) {
+        threads.add(new Thread(
```
Would a thread pool do a better job than manually creating and destroying threads?

jiacheliu3 commented on core/server/worker/src/test/java/alluxio/worker/block/BlockRWLockPoolTest.java:
```diff
@@ -0,0 +1,123 @@
+      }
+      threads.forEach(Thread::start);
+      barrier1.await();
+      Assert.assertEquals(0, pool.mRemainingPoolResources.get());
```
we typically do `import static ...Assert;`

jiacheliu3 commented on core/server/worker/src/test/java/alluxio/worker/block/BlockRWLockPoolTest.java:
```diff
@@ -0,0 +1,123 @@
+      try {
+        Thread.sleep(3000);
+      } catch (Exception e) {
+        System.out.println(e.getMessage());
```
Don't print in tests, you can just throw them as RuntimeException

jiacheliu3 commented on core/server/worker/src/test/java/alluxio/worker/block/BlockRWLockPoolTest.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.concurrent.TimeUnit;
+
+public class BlockRWLockPoolTest {
+  /**
+   * To test the number of remaining resources in the pool. We report this value as a metric.
+   * pool size of 10, 10 threads acquire locks and release early
+   * while another 10 threads acquire locks but release late.
+   * in the first checkpoint, the remaining resource should be 0,
+   * in the second checkpoint,
+   * the remaining resource should be the size of the pool as we returned all locks.
+   */
```
I think the comment here is a bit verbose, could you try to break it up into inline `//` comments in your test code?

jiacheliu3 commented on core/server/worker/src/test/java/alluxio/worker/block/BlockRWLockPoolTest.java:
```diff
@@ -0,0 +1,123 @@
+          throw new RuntimeException(e);
+        }
+      });
+      Assert.assertEquals(poolCapacity, pool.mRemainingPoolResources.get());
```
I'm thinking, you are testing two things:
1. When there are more acquirers than resources, the remaining resources will decrease to 0. 
2. When all acquirers release, the remaining resources go back to capacity.

I feel there's a bit more you can test in this case, like validate the blocking behavior when there's no resources, and vice versa. Also I don't quite get why you used two kinds of threads, did you intend to test sth specific about short/long holding?

jiacheliu3 reviewed (changes requested):
Overall LGTM, I think the pool behavior is good. My comments are mainly regarding styling and the unit test. PTAL, thanks!

jiacheliu3 commented on core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java:
```diff
@@ -0,0 +1,106 @@
+      mRemainingPoolResources.decrementAndGet();
+    } else {
+      SAMPLING_LOG.error(
+          "Lock acquisition failed due to lock pool resources being exhausted. "
+              + "Pool capacity size {}; lock acquisition duration {}{}",
```
```suggestion
          "Failed to acquire a block lock because the block lock pool is full! "
              + "Current pool capacity: {}; lock acquisition duration: {}{}. Tune up {} to increase block lock pool size.",
          mMaxCapacity, time, unit, PropertyKey.WORKER_TIERED_STORE_BLOCK_LOCKS
```

jiacheliu3 reviewed (commented):
Improved the log msg a bit to reflect the property to tune
